# LINEAR REGRESSION

In [1]:
pip install dmba

Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install graphviz

Note: you may need to restart the kernel to use updated packages.


In [3]:
pip install xgboost

Note: you may need to restart the kernel to use updated packages.


In [4]:
%matplotlib inline

from pathlib import Path

import pandas as pd
import numpy as np



from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
import matplotlib.pylab as plt
from dmba import plotDecisionTree, classificationSummary, regressionSummary

no display found. Using non-interactive Agg backend


In [5]:
df = pd.read_csv('Dataset2.csv')

In [6]:
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                   11                    34   0.9978  3.51       0.56   
1                   25                    67   0.9968  3.20       0.68   
2                   15                    54   0.9970  3.26       0.65   
3                   17                    60   0.9980  3.16       0.58   
4                   11                    34   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [7]:
X = df.drop(columns=['quality'])
y = df['quality']

# normalize the X, rename the normlized value as norm_X
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X)
norm_X = pd.DataFrame(scaler.transform(X), columns=X.columns)

In [8]:
norm_X

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0         -0.528360          0.961877    -1.391472       -0.453218  -0.243707   
1         -0.298547          1.967442    -1.391472        0.043416   0.223875   
2         -0.298547          1.297065    -1.186070       -0.169427   0.096353   
3          1.654856         -1.384443     1.484154       -0.453218  -0.264960   
4         -0.528360          0.961877    -1.391472       -0.453218  -0.243707   
...             ...               ...          ...             ...        ...   
1594      -1.217796          0.403229    -0.980669       -0.382271   0.053845   
1595      -1.390155          0.123905    -0.877968       -0.240375  -0.541259   
1596      -1.160343         -0.099554    -0.723916       -0.169427  -0.243707   
1597      -1.390155          0.654620    -0.775267       -0.382271  -0.264960   
1598      -1.332702         -1.216849     1.021999        0.752894  -0.434990   

      free sulfur dioxide  total sulfur dioxide   density        pH  \
0               -0.466240             -0.379145  0.558274  1.288643   
1                0.872555              0.624333  0.028261 -0.719933   
2               -0.083727              0.229023  0.134264 -0.331177   
3                0.107529              0.411474  0.664277 -0.979104   
4               -0.466240             -0.379145  0.558274  1.288643   
...                   ...                   ...       ...       ...   
1594             1.541953             -0.075061 -0.978765  0.899886   
1595             2.211351              0.137798 -0.862162  1.353436   
1596             1.255068             -0.196694 -0.533554  0.705508   
1597             1.541953             -0.075061 -0.676657  1.677400   
1598             0.203158             -0.135878 -0.666057  0.511130   

      sulphates   alcohol  
0     -0.579207 -0.960246  
1      0.128950 -0.584777  
2     -0.048089 -0.584777  
3     -0.461180 -0.584777  
4     -0.579207 -0.960246  
...         ...       ...  
1594  -0.461180  0.072294  
1595   0.601055  0.729364  
1596   0.542042  0.541630  
1597   0.305990 -0.209308  
1598   0.010924  0.541630  

[1599 rows x 11 columns]

In [9]:
train_X, valid_X, train_y, valid_y = train_test_split(norm_X, y, test_size=0.2, random_state=1)

In [10]:
from sklearn.linear_model import LinearRegression

car_lm = LinearRegression()
car_lm.fit(train_X, train_y)

regressionSummary(train_y, car_lm.predict(train_X))
regressionSummary(valid_y, car_lm.predict(valid_X))


Regression statistics

                      Mean Error (ME) : 0.0000
       Root Mean Squared Error (RMSE) : 0.6527
            Mean Absolute Error (MAE) : 0.5077
          Mean Percentage Error (MPE) : -1.4400
Mean Absolute Percentage Error (MAPE) : 9.3009

Regression statistics

                      Mean Error (ME) : -0.0562
       Root Mean Squared Error (RMSE) : 0.6189
            Mean Absolute Error (MAE) : 0.4748
          Mean Percentage Error (MPE) : -2.2414
Mean Absolute Percentage Error (MAPE) : 8.7961


# In Linear Regression model, the training and testing data RMSE are nearby being 0.65 and 0.61 respectively. For model evaluation purposes, we would consider testing data RMSE which is 0.61. The prediction in this model is taking both dependent and independent variables for evaluating the model.

# RANDOM FOREST

In [16]:
from sklearn.ensemble import RandomForestRegressor

rftree = RandomForestRegressor(n_estimators=600)
rftree.fit(train_X, train_y)

RandomForestRegressor(n_estimators=600)

In [17]:
regressionSummary(train_y, rftree.predict(train_X))
regressionSummary(valid_y, rftree.predict(valid_X))


Regression statistics

                      Mean Error (ME) : -0.0077
       Root Mean Squared Error (RMSE) : 0.2114
            Mean Absolute Error (MAE) : 0.1537
          Mean Percentage Error (MPE) : -0.5748
Mean Absolute Percentage Error (MAPE) : 2.8428

Regression statistics

                      Mean Error (ME) : -0.0600
       Root Mean Squared Error (RMSE) : 0.5645
            Mean Absolute Error (MAE) : 0.4041
          Mean Percentage Error (MPE) : -2.1264
Mean Absolute Percentage Error (MAPE) : 7.5395


# # In Random Forest model, the training data RMSE is 0.21 and testing data RMSE is 0.56 even after playing around with estimators from 100 to 1000. For model evaluation purposes, we would consider testing data RMSE which is 0.56. Random Forest model is taking 600 estimators and is building decision tree with different samples and is taking majority vote for classification and average in regression.

# KNN

In [13]:
from sklearn.neighbors import KNeighborsRegressor
knntree = KNeighborsRegressor(n_neighbors = 3, weights = 'distance')
knntree.fit(train_X, train_y)
regressionSummary(train_y, knntree.predict(train_X))
regressionSummary(valid_y, knntree.predict(valid_X))


Regression statistics

                      Mean Error (ME) : 0.0000
       Root Mean Squared Error (RMSE) : 0.0000
            Mean Absolute Error (MAE) : 0.0000
          Mean Percentage Error (MPE) : 0.0000
Mean Absolute Percentage Error (MAPE) : 0.0000

Regression statistics

                      Mean Error (ME) : -0.0467
       Root Mean Squared Error (RMSE) : 0.6316
            Mean Absolute Error (MAE) : 0.3979
          Mean Percentage Error (MPE) : -1.8613
Mean Absolute Percentage Error (MAPE) : 7.4906


In [20]:
# scan through differnt k, to find the best K

import math
from sklearn.metrics import mean_squared_error
results = []
for k in range(1, 8):
    knn = KNeighborsRegressor(n_neighbors=k,weights = 'distance' ).fit(train_X, train_y)
    results.append({
        'k': k,
        'RMSE': math.sqrt(mean_squared_error(valid_y, knn.predict(valid_X)))
    })

# Convert results to a pandas data frame
results = pd.DataFrame(results)
results

k      RMSE
0  1  0.752080
1  2  0.658691
2  3  0.631616
3  4  0.628302
4  5  0.615803
5  6  0.608109
6  7  0.599460

# # In KNN model, the RMSE for the nearest neighbour with K=7 is around 0.599 which is approximately around 0.60. Therefore, the quality of prediction here is somewhat dependent on the distant measure.

# XGBOOST

In [15]:

from xgboost import XGBRegressor
model = XGBRegressor()

model.fit(train_X, train_y)

regressionSummary(train_y, model.predict(train_X))
regressionSummary(valid_y, model.predict(valid_X))


Regression statistics

                      Mean Error (ME) : 0.0000
       Root Mean Squared Error (RMSE) : 0.0376
            Mean Absolute Error (MAE) : 0.0241
          Mean Percentage Error (MPE) : -0.0378
Mean Absolute Percentage Error (MAPE) : 0.4380

Regression statistics

                      Mean Error (ME) : -0.0237
       Root Mean Squared Error (RMSE) : 0.6047
            Mean Absolute Error (MAE) : 0.4004
          Mean Percentage Error (MPE) : -1.3594
Mean Absolute Percentage Error (MAPE) : 7.3736


# In XGBoost model, the RMSE of training data is 0.03 and training data is 0.60. We would consider RMSE of testing data for evaluation. XGBoost model is tuning multiple hyper-parameters with gradient boosting method.

# As the RMSE for Random Forest model is the lowest which is around 0.56, therefore, is shows the data is concentrated around the line of best fit.